srml/session/src/lib.rs

```diff
@@ -332,33 +343,47 @@ decl_module! {
 }
 
 impl<T: Trait> Module<T> {
-	/// Move on to next session: register the new authority set.
+	/// Move on to next session. Register new validator set and session keys. Changes
+	/// to the validator set have a session of delay to take effect. This allows for
+	/// equivocation punishment after a fork.
 	pub fn rotate_session() {
-		// Increment current session index.
 		let session_index = CurrentIndex::get();
 
-		let mut changed = Changed::take();
+		let changed = QueuedChanged::take();
+		let mut next_changed = Changed::take();
 
-		// See if we have a new validator set.
-		let validators = if let Some(new) = T::OnSessionEnding::on_session_ending(session_index) {
-			changed = true;
-			<Validators<T>>::put(&new);
-			new
+		// Get queued session keys and validators.
+		let session_keys = <QueuedKeys<T>>::get();
+		let validators = session_keys.iter()
```

Comment:
If you use `into_iter()` you don't have to use `to_owned()` below.

Reply:
I don't think there's anything to be gained since the session_keys need to be sent to the session handler later

Reply:
Ahh yeah, oversaw this.